server/lib/controller/rule_controller.dart

```diff
@@ -0,0 +1,54 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:aqueduct/aqueduct.dart';
+
+import '../proto/create_rule_request.pb.dart';
+import '../service/firestore.dart';
+import '../service/scheduler.dart';
+import '../utils.dart';
+
+/// A controller that handles operations on rules.
+class RuleController extends ResourceController {
+  /// The client that interacts with Cloud Firestore.
+  final FirestoreClient _firestoreClient;
+
+  /// The client that interacts with Cloud Scheduler.
+  final SchedulerClient _schedulerClient;
+
+  /// Creates an instance of [RuleController].
+  RuleController(this._firestoreClient, this._schedulerClient);
+
+  /// The accepted content types for this controller.
+  @override
+  List<ContentType> acceptedContentTypes = [
+    ContentType('application', 'x-protobuf')
+  ];
+
+  /// Adds the rule to Firestore and schedules it using Scheduler.
+  ///
+  /// Throws [ArgumentError] if request doesn't contain an Authorization header.
+  /// Throws an [ApiRequestError] if Firestore API or Scheduler API returns an
+  /// error. Returns an OK 200 response upon success.
+  @Operation.post()
+  Future<Response> createRule(@Bind.body() List<int> body) async {
+    // The ID token is sent in the `Authorization` header of the request.
+    final authorizationHeader = request.raw.headers.value('Authorization');
+    if (authorizationHeader == null) {
+      throw ArgumentError('Request does not contain an Authorization header');
+    }
+    final idToken = authorizationHeader.split(' ').last;
+
+    final userId = getUserId(idToken);
```

Comment:
this is used in multiple places, maybe consider extract them into util class?

Reply:
Ok, I've created an [issue](https://github.com/googleinterns/dv360-excel-plugin/issues/112), and I'll resolve it after all these controllers are merged in.